src/services/codefixes/fixUnusedIdentifier.ts

```diff
@@ -238,29 +236,26 @@ namespace ts.codefix {
         }
     }
 
-    function mayDeleteParameter(p: ParameterDeclaration, checker: TypeChecker, isFixAll: boolean): boolean {
-        const { parent } = p;
+    function mayDeleteParameter(checker: TypeChecker, sourceFile: SourceFile, parameter: ParameterDeclaration, isFixAll: boolean): boolean {
+        const { parent } = parameter;
         switch (parent.kind) {
             case SyntaxKind.MethodDeclaration:
                 // Don't remove a parameter if this overrides something.
                 const symbol = checker.getSymbolAtLocation(parent.name)!;
                 if (isMemberSymbolInBaseType(symbol, checker)) return false;
                 // falls through
-
             case SyntaxKind.Constructor:

```

Comment:
don't constructors have the same constraints as functions? I don't understand why the new restriction only applies to named functions used "like a callback".

Reply:
```ts
declare function fn(cb: (x: number, y: string) => void): void;
```

```ts
fn((a, b) => b);
```
`a` will not be removed, because there is a condition to handle that. however, this condition is not applying to `FunctionDeclaration` which can be used as a callback too.

https://github.com/microsoft/TypeScript/blob/2b0278a88f766e87bd0fb839b1048d1e17894969/src/services/codefixes/fixUnusedIdentifier.ts#L254-L263

```ts
function callback(a: number, b: number) { b; };
fn(callback); 
```
`a` will be removed because when the QF checks the possibility to remove parameters for `FunctionDeclaration` it 
always allows that 

https://github.com/microsoft/TypeScript/blob/2b0278a88f766e87bd0fb839b1048d1e17894969/src/services/codefixes/fixUnusedIdentifier.ts#L251-L252

as a result,  

```ts
function callback(b: number) { b; };
fn(callback); 
```
However, deleting `a` has to be disallowed for that case. About constructors, not sure, maybe there are some cases where we need to apply the same logic as for `FunctionDeclaration`.

Reply:
ok, that makes sense. I think that eventually we'll want to expand the restrictions, or make parameter deletion smarter.